src/solvers/cg_solver.hpp

```diff
@@ -0,0 +1,282 @@
+
+    auto sync = itl.AddTask(TaskQualifier::local_sync, none,
+                            []() { return TaskStatus::complete; });
+    auto reset = itl.AddTask(
+        TaskQualifier::once_per_region, sync, "update values",
+        [](CGSolver *solver) {
+          solver->ru_old = solver->ru.val;
+          solver->iter_counter++;
+          return TaskStatus::complete;
+        },
+        this);
+
+    // 1. u <- M r
+    auto precon = reset;
+    if (params_.precondition) {
+      auto set_rhs = itl.AddTask(precon, TF(CopyData<FieldTL>), md_r, md_rhs);
+      auto zero_u = itl.AddTask(precon, TF(SetToZero<FieldTL>), md_u);
+      precon =
+          preconditioner.AddLinearOperatorTasks(itl, set_rhs | zero_u, partition, pmesh);
+    } else {
+      precon = itl.AddTask(precon, TF(CopyData<FieldTL>), md_r, md_u);
+    }
+
+    // 2. beta <- r dot u / r dot u {old}
+    auto get_ru = DotProduct<FieldTL>(precon, itl, &ru, md_r, md_u);
+
+    // 3. p <- u + beta p
+    auto correct_p = itl.AddTask(
+        get_ru, "p <- u + beta p",
+        [](CGSolver *solver, std::shared_ptr<MeshData<Real>> &md_u,
+           std::shared_ptr<MeshData<Real>> &md_p) {
+          Real beta = solver->iter_counter > 0 ? solver->ru.val / solver->ru_old : 0.0;
+          return AddFieldsAndStore<FieldTL>(md_u, md_p, md_p, 1.0, beta);
+        },
+        this, md_u, md_p);
+
+    // 4. v <- A p
+    auto comm =
+        AddBoundaryExchangeTasks<BoundaryType::any>(correct_p, itl, md_p, multilevel);
+    auto get_v = eqs_.Ax(itl, comm, md_base, md_p, md_v);
+
+    // 5. alpha <- r dot u / p dot v (calculate denominator)
+    auto get_pAp = DotProduct<FieldTL>(get_v, itl, &pAp, md_p, md_v);
+
+    // 6. x <- x + alpha p
+    auto correct_x = itl.AddTask(
+        get_pAp, "x <- x + alpha p",
+        [](CGSolver *solver, std::shared_ptr<MeshData<Real>> &md_x,
+           std::shared_ptr<MeshData<Real>> &md_p) {
+          Real alpha = solver->ru.val / solver->pAp.val;
+          return AddFieldsAndStore<FieldTL>(md_x, md_p, md_x, 1.0, alpha);
+        },
+        this, md_x, md_p);
+
+    // 6. r <- r - alpha A p
+    auto correct_r = itl.AddTask(
+        get_pAp, "r <- r - alpha A p",
+        [](CGSolver *solver, std::shared_ptr<MeshData<Real>> &md_r,
+           std::shared_ptr<MeshData<Real>> &md_v) {
+          Real alpha = solver->ru.val / solver->pAp.val;
+          return AddFieldsAndStore<FieldTL>(md_r, md_v, md_r, 1.0, -alpha);
+        },
+        this, md_r, md_v);
+
+    // 7. Check and print out residual
+    auto get_res = DotProduct<FieldTL>(correct_r, itl, &residual, md_r, md_r);
+
+    auto print = itl.AddTask(
+        TaskQualifier::once_per_region, get_res,
+        [&](CGSolver *solver, Mesh *pmesh) {
+          Real rms_res = std::sqrt(solver->residual.val / pmesh->GetTotalCells());
+          if (Globals::my_rank == 0 && solver->params_.print_per_step)
+            printf("%i %e\n", solver->iter_counter, rms_res);
```

Comment:
I find reading output like this can be difficult. I like to add tab offsets.
```suggestion
           printf("\t%i %e\n", solver->iter_counter, rms_res);
```